dragonfly/implementations/psyllid_provider.py

```diff
@@ -233,51 +235,181 @@ def get_number_of_streams(self, channel):
-        request = '.node-config.ch'+str(self.channel_dict[channel])+'.fmt.threshold-power-snr'
-        result = self.provider.set(self.queue_dict[channel]+request, snr)
-        self.deactivate(channel)
-        time.sleep(1)
-        self.activate(channel)
+
+        self._set_fmt_snr_threshold( threshold, channel)
+        if threshold_high > threshold:
+            self._set_fmt_snr_high_threshold( threshold_high, channel)
+            self._set_trigger_mode( "two-level-trigger", channel)
+        else:
+            self._set_trigger_mode( "single-level-trigger", channel)
+
+        result = self._set_n_triggers( n_triggers, channel)
+
+        if result != True:
+            return False
+
+        return True
+
+
+    def get_trigger_configuration(self, channel='a'):
+        if self.mode_dict[channel] != 'triggered':
+            logger.warning('Psyllid not in triggered mode')
+            return {'threshold': False, 'threshold_high' : False, 'n_triggers' : False}
+
+        threshold = self._get_fmt_snr_threshold( channel)
+        threshold_high = self._get_fmt_snr_high_threshold( channel)
+        n_triggers = self._get_n_triggers( channel)
+        trigger_mode = self._get_trigger_mode( channel )
+
+        # threshold_high !> threhold results in trigger_mode = 1
+        # threshold_high is not used in this case
+        if trigger_mode == 1:
+            threshold_high = None
+
+        return {'threshold': threshold, 'threshold_high' : threshold_high, 'n_triggers' : n_triggers}
+
+
+    # pre-trigger and skip-tolerance control
+    def set_time_window(self, pretrigger_time=2e-3, skip_tolerance=5e-3, channel='a'):
+        if self.mode_dict[channel] != 'triggered':
+            logger.warning('Psyllid not in triggered mode')
+            return False
+
+        self._set_pretrigger_time( pretrigger_time, channel)
+        result = self._set_skip_tolerance( skip_tolerance, channel)
+
+        if result != True:
+            return False
+
+        self.reactivate(channel)
+        return True
+
+
+    def get_time_window(self, channel='a'):
+        if self.mode_dict[channel] != 'triggered':
+            logger.warning('Psyllid not in triggered mode')
+            return False
+
+        pretrigger_time = self._get_pretrigger_time( channel)
+        skip_tolerance = self._get_skip_tolerance( channel)
+
+        return {'pretrigger_time': pretrigger_time, 'skip_tolerance': skip_tolerance}
+
+
+    # internal psyllid time window and trigger parameter sets and gets
+    def _set_pretrigger_time(self, pretrigger_time, channel='a'):
+        n_pretrigger_packets = int(round(pretrigger_time/4.096e-5))
+        logger.info('Setting psyllid pretrigger to {} packets'.format(n_pretrigger_packets))
+        request = '.node-config.{}.eb.pretrigger'.format(str(self.channel_dict[channel]))
+        self.provider.set(self.queue_dict[channel]+request, n_pretrigger_packets)
         return True
 
 
+    def _get_pretrigger_time(self, channel='a'):
```

Comment:
is there a reason that these get/set methods are all prefixed by `_`? That usually indicates that they are intended for internal use only. it isn't obvious to me why these couldn't be used externally, though maybe that is indeed the intent

Reply:
They could be used externally. Initially they were not prefixed by _ and I did use them to set and get trigger settings. But as the number of configurable parameters grew I became lazy and added set_trigger_configuration (sets 4 trigger parameters) and set_time_window (sets pretrigger and skip tolerance). I only use these now and the roach_daq_interface also only uses these. Therefore I added the prefixes to make them internal.
In execution scripts it probably doesn't matter whether you have 6 sets or 2. You set 6 parameters either way. Would you prefer to have the 6 set methods for external use?